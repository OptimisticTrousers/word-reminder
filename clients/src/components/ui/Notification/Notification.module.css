.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --text-color: var(--success-text-color);
  --icon-color: var(--success-icon-color);
  --button-bkg-color: var(--success-button-bkg-color);
  --secondary-button-bkg-color: var(--secondary-success-button-bkg-color);
  background-color: var(--success-bkg-color);
}

.notification--error {
  --text-color: var(--alert-text-color);
  --icon-color: var(--alert-icon-color);
  --button-bkg-color: var(--alert-button-bkg-color);
  --secondary-button-bkg-color: var(--secondary-alert-button-bkg-color);
  background-color: var(--alert-bkg-color);
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--text-color);
  padding: 0.375rem;
  background-color: var(--button-bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__button:hover {
  background-color: var(--secondary-button-bkg-color);
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
