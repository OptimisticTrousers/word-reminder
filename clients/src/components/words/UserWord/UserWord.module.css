.user-word__word,
.user-word__panel {
  margin-top: 1rem;
}

.user-word__panel {
  border: 1px solid var(--border-color);
}

.word {
  background-color: var(--secondary-card-bkg-color);
  padding: 1rem;
  border-radius: 8px;
  border: 1px solid var(--border-color);
  transition: 0.4s;
}

.user-word__header {
  display: flex;
  justify-content: space-between;
  margin-bottom: 0.5rem;
}

.word__buttons {
  display: flex;
  align-items: center;
  gap: 0.75rem;
  margin-top: 0.75rem;
}

.word__buttons > * {
  flex: 1;
}

.word__button {
  display: flex;
  align-items: center;
  justify-content: center;
}

.word__button .word__icon {
  margin-right: 8px;
  width: 28px;
  height: 28px;
}

.word__form {
  display: flex;
  justify-content: center;
}

.word__form > * {
  flex: 1;
}

.user-word__panel {
  padding: 1rem;
  background-color: var(--accordion-bkg-color);
}
