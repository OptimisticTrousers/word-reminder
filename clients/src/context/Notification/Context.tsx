import { createContext } from "react";

import { NotificationContext as Context } from "./Provider";

export enum NOTIFICATION_ACTIONS {
  SUCCESS = "success",
  ERROR = "error",
}

export const NotificationContext = createContext<Context>({
  showNotification: function () {},
  dismissNotification: function () {},
});
