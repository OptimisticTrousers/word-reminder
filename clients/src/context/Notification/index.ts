export { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
export { NotificationProvider } from "./Provider";
