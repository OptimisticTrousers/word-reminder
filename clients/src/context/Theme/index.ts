export { ThemeContext } from "./Context";
export { ThemeProvider } from "./Provider";
