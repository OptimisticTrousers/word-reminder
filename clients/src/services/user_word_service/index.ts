export { userWordService } from "./user_word_service";
