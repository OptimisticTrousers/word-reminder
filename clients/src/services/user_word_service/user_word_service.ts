import { service, Params } from "../service";

export const userWordService = (function (service) {
  const { get, post, put, remove, VITE_API_DOMAIN } = service;

  function getUserWordList({
    userId,
    params,
  }: {
    userId: string;
    params: Params;
  }) {
    return get({
      url: `${VITE_API_DOMAIN}/users/${userId}/userWords`,
      params,
      options: { credentials: "include" },
    });
  }

  function getUserWord({
    userId,
    userWordId,
  }: {
    userId: string;
    userWordId: string;
  }) {
    return get({
      url: `${VITE_API_DOMAIN}/users/${userId}/userWords/${userWordId}`,
      options: { credentials: "include" },
    });
  }

  function createUserWord({
    userId,
    formData,
  }: {
    userId: string;
    formData: FormData;
  }) {
    return post({
      url: `${VITE_API_DOMAIN}/users/${userId}/userWords`,
      options: {
        body: formData,
        credentials: "include",
      },
    });
  }

  function updateUserWord({
    userId,
    userWordId,
    body,
  }: {
    userId: string;
    userWordId: string;
    body: {
      learned: boolean;
    };
  }) {
    return put({
      url: `${VITE_API_DOMAIN}/users/${userId}/userWords/${userWordId}`,
      options: {
        body: JSON.stringify(body),
        credentials: "include",
        mode: "cors",
        headers: {
          "Content-Type": "application/json",
        },
      },
    });
  }

  function deleteUserWord({
    userId,
    userWordId,
  }: {
    userId: string;
    userWordId: string;
  }) {
    return remove({
      url: `${VITE_API_DOMAIN}/users/${userId}/userWords/${userWordId}`,
      options: { credentials: "include" },
    });
  }

  return {
    getUserWordList,
    getUserWord,
    createUserWord,
    updateUserWord,
    deleteUserWord,
  };
})(service);
