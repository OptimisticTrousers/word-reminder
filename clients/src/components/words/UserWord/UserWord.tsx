import { Detail, User } from "common";
import { GraduationCap, Info, Trash } from "lucide-react";
import { FormEvent, useState } from "react";
import CSSModules from "react-css-modules";
import { Link, useOutletContext } from "react-router-dom";

import { DeleteUserWordModal } from "../../modals/DeleteUserWordModal";
import { CondensedWord } from "../CondensedWord";
import styles from "./UserWord.module.css";
import { useMutation, useQueryClient } from "@tanstack/react-query";
import { userWordService } from "../../../services/user_word_service";
import { useNotificationError } from "../../../hooks/useNotificationError";

export const UserWord = CSSModules(
  function ({
    learned,
    created_at,
    updated_at,
    details,
    id,
  }: {
    learned: boolean;
    created_at: Date;
    updated_at: Date;
    details: Detail[];
    id: number;
  }) {
    const [isAccordionOpen, setIsAccordionOpen] = useState(false);
    const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);

    const queryClient = useQueryClient();
    const { user }: { user: User } = useOutletContext();
    const userId = String(user.id);
    const { showNotificationError } = useNotificationError();

    const { isPending, mutate } = useMutation({
      mutationFn: userWordService.updateUserWord,
      onSuccess: () => {
        queryClient.invalidateQueries({
          queryKey: ["userWords"],
        });
      },
      onError: showNotificationError,
    });

    const toggleAccordion = () => {
      setIsAccordionOpen((prevValue) => !prevValue);
    };

    function toggleDeleteModal() {
      setIsDeleteModalOpen((prevIsDeleteModalOpen) => !prevIsDeleteModalOpen);
    }

    function handleLearned(event: FormEvent<HTMLFormElement>) {
      event.preventDefault();
      const formData = new FormData(event.currentTarget);
      mutate({
        userId,
        userWordId: String(id),
        body: {
          learned: (formData.get("learned") as string) === "true",
        },
      });
    }

    const disabled = isPending;

    return (
      <>
        <div styleName="user-word">
          <article styleName="user-word__word word">
            <div styleName="user-word__header">
              <h3 styleName="word__name">{details[0].word}</h3>
              <Link to={`/userWords/${id}`}>More Word Details</Link>
            </div>
            <p styleName="word__name">
              Created At: {created_at.toLocaleString()}
            </p>
            <p styleName="word__name">
              Updated At: {updated_at.toLocaleString()}
            </p>
            <p styleName="word__name">Learned: {learned ? "Yes" : "No"}</p>
            <div styleName="word__buttons">
              <button
                id="accordion-button"
                type="button"
                styleName="word__button word__button--info"
                onClick={toggleAccordion}
                aria-expanded={isAccordionOpen}
                aria-label="More word details"
                aria-controls="accordion"
              >
                <Info styleName="word__icon" />
              </button>
              <form styleName="word__form" onSubmit={handleLearned}>
                <input
                  styleName="word__input"
                  type="hidden"
                  id="learned"
                  name="learned"
                  required
                  defaultValue={String(!learned)}
                />
                <button
                  styleName="word__button"
                  type="submit"
                  disabled={disabled}
                >
                  <GraduationCap styleName="word__icon" />
                  {disabled ? "Toggling Learned..." : "Toggle Learned"}
                </button>
              </form>
              <button
                styleName="word__delete"
                onClick={toggleDeleteModal}
                aria-haspopup={true}
                aria-label="Open delete user word modal"
              >
                <Trash styleName="word__icon" />
              </button>
            </div>
          </article>
          {isAccordionOpen && (
            <div
              id="accordion"
              styleName={`user-word__panel ${
                isAccordionOpen && "user-word__panel--active"
              }`}
              role="region"
              aria-labelledby="accordion-button"
            >
              <CondensedWord details={details} />
            </div>
          )}
        </div>
        {isDeleteModalOpen && (
          <DeleteUserWordModal
            toggleModal={toggleDeleteModal}
            userWordId={String(id)}
          />
        )}
      </>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
