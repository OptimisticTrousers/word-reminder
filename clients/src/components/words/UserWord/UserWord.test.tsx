import { Detail, UserWord as UserWordI } from "common";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { UserWord } from "./UserWord";
import { createRoutesStub, Outlet } from "react-router-dom";
import { userWordService } from "../../../services/user_word_service";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { NotificationProvider } from "../../../context/Notification";

vi.mock("../CondensedWord/CondensedWord", () => {
  return {
    CondensedWord: function ({ details }: { details: Detail[] }) {
      return (
        <>
          <div data-testid="details">{JSON.stringify(details)}</div>
        </>
      );
    },
  };
});

vi.mock("../../modals/DeleteUserWordModal", () => {
  return {
    DeleteUserWordModal: function () {
      return <div data-testid="modal"></div>;
    },
  };
});

const props = {
  id: 1,
  word_id: 1,
  user_id: 1,
  learned: false,
  updated_at: new Date("December 17, 1995 03:24:00"),
  created_at: new Date("December 17, 1995 03:24:00"),
  details: [{ word: "exemplary", phonetics: [], meanings: [] }],
};

describe("UserWord component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  const testUser = {
    id: "1",
  };

  function setup({
    userWord,
    queryClient,
  }: {
    userWord: UserWordI & { details: Detail[] };
    queryClient: QueryClient;
  }) {
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <NotificationProvider>
                  <QueryClientProvider client={queryClient}>
                    <UserWord {...userWord} />
                  </QueryClientProvider>
                </NotificationProvider>
              );
            },
          },
        ],
      },
    ]);

    return {
      user: userEvent.setup(),
      ...render(<Stub initialEntries={["/"]} />),
    };
  }

  it("renders user word component when learned is true", async () => {
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    setup({ userWord: { ...props, learned: true }, queryClient });

    const word = screen.getByText(`${props.details[0].word}`);
    const createdAt = screen.getByText(
      `Created At: ${props.created_at.toLocaleString()}`
    );
    const updatedAt = screen.getByText(
      `Updated At: ${props.created_at.toLocaleString()}`
    );
    const learned = screen.getByText("Learned: Yes");
    const infoButton = screen.getByRole("button", {
      name: "More word details",
    });
    const toggleLearnedButton = screen.getByRole("button", {
      name: "Toggle Learned",
    });
    const deleteButton = screen.getByRole("button", {
      name: "Open delete user word modal",
    });
    const moreWordDetailsLink = screen.getByRole("link", {
      name: "More Word Details",
    });
    expect(word).toBeInTheDocument();
    expect(createdAt).toBeInTheDocument();
    expect(updatedAt).toBeInTheDocument();
    expect(learned).toBeInTheDocument();
    expect(infoButton).toBeInTheDocument();
    expect(toggleLearnedButton).toBeInTheDocument();
    expect(deleteButton).toBeInTheDocument();
    expect(moreWordDetailsLink).toBeInTheDocument();
  });

  it("renders user word component when learned is false", async () => {
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    setup({ userWord: { ...props, learned: false }, queryClient });

    const word = screen.getByText(`${props.details[0].word}`);
    const createdAt = screen.getByText(
      `Created At: ${props.created_at.toLocaleString()}`
    );
    const updatedAt = screen.getByText(
      `Updated At: ${props.created_at.toLocaleString()}`
    );
    const learned = screen.getByText("Learned: No");
    const infoButton = screen.getByRole("button", {
      name: "More word details",
    });
    const toggleLearnedButton = screen.getByRole("button", {
      name: "Toggle Learned",
    });
    const deleteButton = screen.getByRole("button", {
      name: "Open delete user word modal",
    });
    const moreWordDetailsLink = screen.getByRole("link", {
      name: "More Word Details",
    });
    expect(word).toBeInTheDocument();
    expect(createdAt).toBeInTheDocument();
    expect(updatedAt).toBeInTheDocument();
    expect(learned).toBeInTheDocument();
    expect(infoButton).toBeInTheDocument();
    expect(toggleLearnedButton).toBeInTheDocument();
    expect(deleteButton).toBeInTheDocument();
    expect(moreWordDetailsLink).toBeInTheDocument();
  });

  describe("when toggling learned property", () => {
    const status = 200;

    it("calls the functions to toggle learned when user word is currently false", async () => {
      const queryClient = new QueryClient({
        defaultOptions: { queries: { retry: false } },
      });
      const mockUpdateUserWord = vi
        .spyOn(userWordService, "updateUserWord")
        .mockImplementation(async () => {
          return {
            json: { userWord: props },
            status,
          };
        });
      const mockInvalidateQueries = vi
        .spyOn(queryClient, "invalidateQueries")
        .mockImplementation(vi.fn());
      const body = {
        learned: true,
      };
      const { user } = setup({
        userWord: { ...props, learned: false },
        queryClient,
      });

      const toggleLearnedButton = screen.getByRole("button", {
        name: "Toggle Learned",
      });
      await user.click(toggleLearnedButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(mockUpdateUserWord).toHaveBeenCalledTimes(1);
      expect(mockUpdateUserWord).toHaveBeenCalledWith({
        userId: testUser.id,
        userWordId: String(props.id),
        body,
      });
      expect(mockInvalidateQueries).toHaveBeenCalledTimes(1);
      expect(mockInvalidateQueries).toHaveBeenCalledWith({
        queryKey: ["userWords"],
      });
    });

    it("calls the functions to toggle learned when user word is currently true", async () => {
      const queryClient = new QueryClient({
        defaultOptions: { queries: { retry: false } },
      });
      const mockUpdateUserWord = vi
        .spyOn(userWordService, "updateUserWord")
        .mockImplementation(async () => {
          return {
            json: { userWord: props },
            status,
          };
        });
      const mockInvalidateQueries = vi
        .spyOn(queryClient, "invalidateQueries")
        .mockImplementation(vi.fn());
      const body = {
        learned: false,
      };
      const { user } = setup({
        userWord: { ...props, learned: true },
        queryClient,
      });

      const toggleLearnedButton = screen.getByRole("button", {
        name: "Toggle Learned",
      });
      await user.click(toggleLearnedButton);

      const notification = screen.queryByRole("dialog");
      expect(toggleLearnedButton).not.toBeDisabled();
      expect(notification).not.toBeInTheDocument();
      expect(mockUpdateUserWord).toHaveBeenCalledTimes(1);
      expect(mockUpdateUserWord).toHaveBeenCalledWith({
        userId: testUser.id,
        userWordId: String(props.id),
        body,
      });
      expect(mockInvalidateQueries).toHaveBeenCalledTimes(1);
      expect(mockInvalidateQueries).toHaveBeenCalledWith({
        queryKey: ["userWords"],
      });
    });

    it("calls the functions to show a notification error", async () => {
      const message = "Bad Request.";
      const queryClient = new QueryClient({
        defaultOptions: { queries: { retry: false } },
      });
      const mockUpdateUserWord = vi
        .spyOn(userWordService, "updateUserWord")
        .mockImplementation(async () => {
          return Promise.reject({
            json: { message },
            status: 400,
          });
        });
      const mockInvalidateQueries = vi
        .spyOn(queryClient, "invalidateQueries")
        .mockImplementation(vi.fn());
      const body = {
        learned: !props.learned,
      };
      const { user } = setup({ userWord: props, queryClient });

      const toggleLearnedButton = screen.getByRole("button", {
        name: "Toggle Learned",
      });
      await user.click(toggleLearnedButton);

      const notification = screen.getByRole("dialog", { name: message });
      expect(notification).toBeInTheDocument();
      expect(toggleLearnedButton).not.toBeDisabled();
      expect(mockUpdateUserWord).toHaveBeenCalledTimes(1);
      expect(mockUpdateUserWord).toHaveBeenCalledWith({
        userId: testUser.id,
        userWordId: String(props.id),
        body,
      });
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });

    it("disables the toggle learned button when the mutation is loading", async () => {
      const delay = 50;
      const queryClient = new QueryClient({
        defaultOptions: { queries: { retry: false } },
      });
      const mockUpdateUserWord = vi
        .spyOn(userWordService, "updateUserWord")
        .mockImplementation(async () => {
          return new Promise((resolve) => {
            setTimeout(() => {
              resolve({
                json: { userWord: props },
                status,
              });
            }, delay);
          });
        });
      const mockInvalidateQueries = vi
        .spyOn(queryClient, "invalidateQueries")
        .mockImplementation(vi.fn());
      const body = {
        learned: !props.learned,
      };
      const { user } = setup({ userWord: props, queryClient });

      const toggleLearnedButton = screen.getByRole("button", {
        name: "Toggle Learned",
      });
      await user.click(toggleLearnedButton);

      const notification = screen.queryByRole("dialog");
      expect(notification).not.toBeInTheDocument();
      expect(toggleLearnedButton).toBeDisabled();
      expect(toggleLearnedButton).toHaveTextContent("Toggling Learned...");
      expect(mockUpdateUserWord).toHaveBeenCalledTimes(1);
      expect(mockUpdateUserWord).toHaveBeenCalledWith({
        userId: testUser.id,
        userWordId: String(props.id),
        body,
      });
      expect(mockInvalidateQueries).not.toHaveBeenCalled();
    });
  });

  it("does not renders the word by default", async () => {
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    setup({ userWord: props, queryClient });

    const word = screen.getByText("exemplary");
    const createdAt = screen.getByText(
      `Created At: ${props.created_at.toLocaleString()}`
    );
    const updatedAt = screen.getByText(
      `Updated At: ${props.updated_at.toLocaleString()}`
    );
    const details = screen.queryByTestId("details");
    expect(word).toBeInTheDocument();
    expect(createdAt).toBeInTheDocument();
    expect(updatedAt).toBeInTheDocument();
    expect(details).not.toBeInTheDocument();
  });

  it("it opens delete modal", async () => {
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const { user } = setup({ userWord: props, queryClient });

    const deleteButton = screen.getByRole("button", {
      name: "Open delete user word modal",
    });
    await user.click(deleteButton);

    const deleteModal = screen.getByTestId("modal");
    expect(deleteModal).toBeInTheDocument();
  });

  it("it opens the accordion", async () => {
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const { user, asFragment } = setup({ userWord: props, queryClient });

    const infoButton = screen.getByRole("button", {
      name: "More word details",
    });
    await user.click(infoButton);

    const details = screen.getByTestId("details");
    expect(details).toHaveTextContent(JSON.stringify(props.details));
    const word = screen.getByText("exemplary");
    const createdAt = screen.getByText(
      `Created At: ${props.created_at.toLocaleString()}`
    );
    const updatedAt = screen.getByText(
      `Updated At: ${props.updated_at.toLocaleString()}`
    );
    expect(word).toBeInTheDocument();
    expect(createdAt).toBeInTheDocument();
    expect(updatedAt).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("it closes the accordion", async () => {
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const { user } = setup({ userWord: props, queryClient });

    const infoButton = screen.getByRole("button", {
      name: "More word details",
    });
    await user.click(infoButton);
    await user.click(infoButton);

    const details = screen.queryByTestId("details");
    expect(details).not.toBeInTheDocument();
  });
});
