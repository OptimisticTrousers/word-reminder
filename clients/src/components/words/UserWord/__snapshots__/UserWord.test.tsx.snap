// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserWord component > it opens the accordion 1`] = `
<DocumentFragment>
  <div
    class="_user-word_a41cd7"
  >
    <article
      class="_user-word__word_a41cd7 _word_a41cd7"
    >
      <div
        class="_user-word__header_a41cd7"
      >
        <h3
          class="_word__name_a41cd7"
        >
          exemplary
        </h3>
        <a
          data-discover="true"
          href="/userWords/1"
        >
          More Word Details
        </a>
      </div>
      <p
        class="_word__name_a41cd7"
      >
        Created At: 12/17/1995, 3:24:00 AM
      </p>
      <p
        class="_word__name_a41cd7"
      >
        Updated At: 12/17/1995, 3:24:00 AM
      </p>
      <p
        class="_word__name_a41cd7"
      >
        Learned: No
      </p>
      <div
        class="_word__buttons_a41cd7"
      >
        <button
          aria-controls="accordion"
          aria-expanded="true"
          aria-label="More word details"
          class="_word__button_a41cd7 _word__button--info_a41cd7"
          id="accordion-button"
          type="button"
        >
          <svg
            class="lucide lucide-info _word__icon_a41cd7"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <path
              d="M12 16v-4"
            />
            <path
              d="M12 8h.01"
            />
          </svg>
        </button>
        <form
          class="_word__form_a41cd7"
        >
          <input
            class="_word__input_a41cd7"
            id="learned"
            name="learned"
            required=""
            type="hidden"
            value="true"
          />
          <button
            class="_word__button_a41cd7"
            type="submit"
          >
            <svg
              class="lucide lucide-graduation-cap _word__icon_a41cd7"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21.42 10.922a1 1 0 0 0-.019-1.838L12.83 5.18a2 2 0 0 0-1.66 0L2.6 9.08a1 1 0 0 0 0 1.832l8.57 3.908a2 2 0 0 0 1.66 0z"
              />
              <path
                d="M22 10v6"
              />
              <path
                d="M6 12.5V16a6 3 0 0 0 12 0v-3.5"
              />
            </svg>
            Toggle Learned
          </button>
        </form>
        <button
          aria-haspopup="true"
          aria-label="Open delete user word modal"
          class="_word__delete_a41cd7"
        >
          <svg
            class="lucide lucide-trash _word__icon_a41cd7"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 6h18"
            />
            <path
              d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
            />
            <path
              d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
            />
          </svg>
        </button>
      </div>
    </article>
    <div
      aria-labelledby="accordion-button"
      class="_user-word__panel_a41cd7 _user-word__panel--active_a41cd7"
      id="accordion"
      role="region"
    >
      <div
        data-testid="details"
      >
        [{"word":"exemplary","phonetics":[],"meanings":[]}]
      </div>
    </div>
  </div>
</DocumentFragment>
`;
